File: ERP/xe-toolkit/src/plugins/CaptiveCollector/windows/sysinfo.cpp
/****************************************************************************
**
** Copyright (c) 2006-2007 Captive Metrics Software Corporation.
**                    All rights reserved.
**
** This file is part of the Windows plugin library for CaptiveCollector.
**
** This file may be used under the terms of the GNU General Public
** License version 2.0 as published by the Free Software Foundation
** and appearing in the file LICENSE-GPL.txt contained within the
** same package as this file. This software is subject to a
** dual-licensing mechanism, the details of which are outlined in
** file LICENSE-DUAL.txt, also contained within this package. Be sure
** to use the correct license for your needs. To view the commercial
** license, read LICENSE-COMMERCIAL.txt also contained within this
** package.
**
** If you do not have access to these files or are unsure which license
** is appropriate for your use, please contact the sales department at
** sales@captivemetrics.com.
**
** This file is provided AS IS with NO WARRANTY OF ANY KIND, INCLUDING THE
** WARRANTY OF DESIGN, MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
**
****************************************************************************/


#include "Stdafx.h"
#include "sysinfo.h"
#include "plugin_ex.h"
#include <time.h>

sysinfo_t::sysinfo_t()
{
    SYSTEM_INFO si;
    DWORD error;
    HKEY key;
    int i;
    int online = 0;

    GetSystemInfo(&si);
    pageSize = (int) si.dwPageSize;
    cpuCountConfigured = (int) si.dwNumberOfProcessors;

    for (i = 0; i < 32; i++)
        if (si.dwActiveProcessorMask & (1 << i))
            online++;

    cpuCountOnline = online;

    clockTicksPerSecond = CLOCKS_PER_SEC;

    LPCSTR regent = "HARDWARE\\DESCRIPTION\\System\\CentralProcessor\\0";

    error =
        RegOpenKeyEx(HKEY_LOCAL_MACHINE, regent, 0, KEY_QUERY_VALUE, &key);

    if (error != ERROR_SUCCESS)
    {
        throw plugin_exception("sysinfo_t::sysinfo_t: RegOpenKeyEx fails");
    }

    char identifier[BUFSIZ];
    void *p =
        read_value(key, "Identifier", REG_SZ, identifier, sizeof identifier);
    if (p == 0)
    {
        RegCloseKey(key);
        throw
            plugin_exception("sysinfo_t::sysinfo_t: cannot read Identifier");
    }

    trimout(p, sizeof identifier);
    cpuArchitecture.assign((const char *) p);

    char processorName[BUFSIZ];
    p = read_value(key, "ProcessorNameString", REG_SZ,
                   processorName, sizeof processorName);
    if (p == 0)
    {
        RegCloseKey(key);
        throw plugin_exception(
            "sysinfo_t::sysinfo_t: cannot read ProcessorNameString");
    }

    trimout(p, sizeof processorName);
    cpuModel.assign((const char *) p);

    char MHz[BUFSIZ];
    p = read_value(key, "~MHz", REG_DWORD, MHz, sizeof MHz);
    if (p == 0)
    {
        RegCloseKey(key);
        throw plugin_exception("sysinfo_t::sysinfo_t: cannot read ~MHz");
    }

    cpuClockSpeed = (int) *((DWORD *) p);

    RegCloseKey(key);


    regent = "SOFTWARE\\Microsoft\\Windows NT\\CurrentVersion";

    error =
        RegOpenKeyEx(HKEY_LOCAL_MACHINE, regent, 0, KEY_QUERY_VALUE, &key);

    if (error != ERROR_SUCCESS)
    {
        throw plugin_exception("sysinfo_t::sysinfo_t: RegOpenKeyEx fails");
    }

    char productName[BUFSIZ];
    p = read_value(key, "ProductName", REG_SZ, productName,
                   sizeof productName);
    if (p == 0)
    {
        RegCloseKey(key);
        throw plugin_exception(
            "sysinfo_t::sysinfo_t: cannot read ProductName");
    }
    trimout(p, sizeof productName);
    osPlatform.assign((const char *) p);

    char currentVersion[BUFSIZ];
    p = read_value(key, "CurrentVersion", REG_SZ,
        currentVersion, sizeof currentVersion);
    if (p == 0)
    {
        RegCloseKey(key);
        throw plugin_exception(
            "sysinfo_t::sysinfo_t: cannot read CurrentVersion");
    }
    trimout(p, sizeof currentVersion);
    osVersion.assign((const char *) p);

    RegCloseKey(key);

    MEMORYSTATUS gms;

    GlobalMemoryStatus(&gms);
    MBytesPagingFileTotal =
        (int) ((gms.dwTotalPageFile - gms.dwTotalPhys) / (1024 * 1024));

    HMODULE hModule = LoadLibrary("ntdll.dll");
    if (hModule == 0)
    {
        throw plugin_exception("sysinfo_t::sysinfo_t: cannot attach ntdll.dll");
    }

    FARPROC nqsi = GetProcAddress(hModule, "NtQuerySystemInformation");
    if (nqsi == 0)
    {
        throw plugin_exception(
            "sysinfo_t::sysinfo_t: cannot attach NtQuerySystemInformation");
    }

    typedef NTSTATUS(NTAPI *NQSI)(
        IN SYSTEM_INFORMATION_CLASS SystemInformationClass,
        OUT PVOID SystemInformation,
        IN ULONG SystemInformationLength,
        OUT PULONG ReturnLength OPTIONAL);

    NQSI func = (NQSI) nqsi;

    SYSTEM_BASIC_INFORMATION sbi;

#define SBI_PHYSPAGES(s) (((ULONG *) &(s))[3])

    (*func) (SystemBasicInformation, &sbi, sizeof sbi, 0);

    double physp = (double) SBI_PHYSPAGES(sbi);
    physp *= (double) si.dwPageSize;

    MBytesMemoryTotal = (int) (ULONG) (physp / (1024 * 1024));
}

void sysinfo_t::refresh(void)
{
    SYSTEM_INFO si;
    int i;
    int online = 0;
    MEMORYSTATUS gms;

    GetSystemInfo(&si);
    for (i = 0; i < 32; i++)
        if (si.dwActiveProcessorMask & (1 << i))
            online++;

    cpuCountOnline = online;

    GlobalMemoryStatus(&gms);
    MBytesPagingFileTotal =
        (int) ((gms.dwTotalPageFile - gms.dwTotalPhys) / (1024 * 1024));
}

void *sysinfo_t::read_value(HKEY key, LPCSTR name, DWORD type, char buf[],
                            DWORD sz)
{
    DWORD error;
    DWORD returnType;

    error = RegQueryValueEx(key, name, 0, &returnType, (BYTE *) buf, &sz);

    if (error != ERROR_SUCCESS && error != ERROR_MORE_DATA)
        return 0;

    if (returnType != type)
        return 0;

    return (void *) buf;
}

void sysinfo_t::trimout(void *s, DWORD sz)
{
    char *p;
    char *q;
    bool altered = false;

    for (p = (char *) s; *p; p++)
        if (*p == ' ' || *p == '\t')
        {
            altered = true;
            p++;
        }
        else
            break;
    q = p;
    p += strlen(p) - 1;
    while (*p == ' ' || *p == '\t')
    {
        altered = true;
        p--;
    }
    *(p + 1) = '\0';

    if (altered)
        strcpy_s((char *) s, sz, q);
}

File: ERP/xe-toolkit/src/java/XE/com/capmet/tools/XE/programs/SysInfo.java
/****************************************************************************
**
** Copyright (c) 2006-2007 Captive Metrics Software Corporation.
**                    All rights reserved.
**
** This file is part of the XE programs for The XE Toolkit.
**
** This file may be used under the terms of the GNU General Public
** License version 2.0 as published by the Free Software Foundation
** and appearing in the file LICENSE-GPL.txt contained within the
** same package as this file. This software is subject to a
** dual-licensing mechanism, the details of which are outlined in
** file LICENSE-DUAL.txt, also contained within this package. Be sure
** to use the correct license for your needs. To view the commercial
** license, read LICENSE-COMMERCIAL.txt also contained within this
** package.
**
** If you do not have access to these files or are unsure which license
** is appropriate for your use, please contact the sales department at
** sales@captivemetrics.com.
**
** This file is provided AS IS with NO WARRANTY OF ANY KIND, INCLUDING THE
** WARRANTY OF DESIGN, MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
**
****************************************************************************/

package com.capmet.tools.XE.programs;

import com.capmet.metrics.cm.CmSysInfo;
import com.capmet.metrics.cm.CmExceptionParser;
import com.capmet.tools.XE.GetOpt;
import com.capmet.tools.XE.XEMessages;

/**
 * An application for displaying information about a host.<p>
 *
 * Use: java SysInfo [-h remote-host]
 */
public class SysInfo
{
    // usage blurb
    private static void usage()
    {
        System.out.println(XEMessages.msgs.getString("SysInfo_Usage")); 
    }

    private static String uptime(int total)
    {
        return total + ": (" + CmSysInfo.formatUpTime(total) + ")";
    }

    /**
     * The main method of the SysInfo application.
     * @param args Arguments passed from the environment
     */
    public static void main(String[] args)
    {
        try {
            String remoteHost = null;
            GetOpt opts = new GetOpt(args, "h:");
            int c;

            opts.optErr = false;
            while((c = opts.getopt()) != opts.optEOF) {
                switch(c) {
                case 'h':
                    remoteHost = opts.getOptArg();
                    break;
                default:
                    usage();
                    return;
                }
            }

            CmSysInfo si = new CmSysInfo(remoteHost);
            String[] tags = new String[] {
                XEMessages.msgs.getString("PF_Free"),
                XEMessages.msgs.getString("Mem_Free"),
                XEMessages.msgs.getString("PF_Size"),
                XEMessages.msgs.getString("Mem_Size"),
                XEMessages.msgs.getString("Clock_Tick"),
                XEMessages.msgs.getString("OS_Platform"),
                XEMessages.msgs.getString("OS_Version"),
                XEMessages.msgs.getString("Page_Size"),
                XEMessages.msgs.getString("CPU_Architect"),
                XEMessages.msgs.getString("CPU_ModelName"),
                XEMessages.msgs.getString("CPUs_Conf"),
                XEMessages.msgs.getString("CPUs_Online"),
                XEMessages.msgs.getString("CPU_Clock"),
                XEMessages.msgs.getString("Uptime")
            };
            Object[] values = new Object[] {
                si.pagingFileFreePercent,
                si.memoryFreePercent,
                si.pagingFileTotalMB,
                si.memoryTotalMB,
                si.clockTicksPerSecond,
                si.osPlatform,
                si.osVersion,
                si.pageSize,
                si.cpuArchitecture,
                si.cpuModel,
                si.cpuCountConfigured,
                si.cpuCountOnline,
                si.cpuClockSpeed,
                uptime(si.systemUpTime.intValue())
            };
            int length = 0;

            // find the lhs with the greatest length
            for(String s : tags) {
                int l = s.length();

                if (l > length)
                    length = l;
            }

            // construct a format string (%-X.Xs = %s\n)
            String format = "%-" + length + "." + length + "s = %s\n";

            for(int i=0; i<tags.length; i++)
                System.out.printf(format, tags[i], values[i].toString());
        } catch(Exception e) {
            System.err.println(new CmExceptionParser(e).getMessage());
        }
    }
}
